File: ts/src/proto.ts
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// @tabnas/proto — a Tabnas plugin that parses Protocol Buffers `.proto`
// IDL (proto2, proto3, edition 2023/2024) into FileDescriptorProto-shaped
// JSON. The version is auto-detected from the file's `syntax`/`edition`
// declaration and/or supplied via the `version` option.

import { Tabnas } from '@tabnas/parser'
import { abnf } from '@tabnas/abnf'

import { grammarText } from './grammar'
import { buildFile } from './build-descriptor'
import { recordAggregate } from './aggregate'
import {
  ProtoVersion, declaredVersion, resolveVersion,
} from './detect-version'
import { FileDescriptorProto } from './descriptor'

export interface ProtoOptions {
  // Explicit protobuf version. When null, auto-detect from the file's
  // syntax/edition declaration.
  version: null | ProtoVersion
  // When true (default), error if an explicit version disagrees with the
  // file's declaration; when false the declaration wins.
  reconcile: boolean
}

type AnyTabnas = Tabnas & { abnf?: Function }

// The Tabnas plugin: installs the union proto grammar so the engine can
// parse `.proto` source into a `{rule, src, kids}` CST. Use the exported
// `parse()` / `toDescriptor()` to turn that CST into a FileDescriptorProto.
const Proto = ((tn: AnyTabnas, _options?: Partial<ProtoOptions>) => {
  // Proto drives parsing through @tabnas/abnf; ensure it is installed.
  if ('function' !== typeof tn.abnf) tn.use(abnf)
  // The grammar admits every keyword as an identifier, which a compiler
  // without `tokenClasses` expands into millions of alternates
  // (tabnas/bnf#71) and never finishes installing. Probe the installed
  // compiler with a three-token grammar first, so an old @tabnas/bnf
  // fails here, at once and by name, rather than hanging.
  const probe = (tn.abnf as any).toSpec('s = a a\na = "x" / "y"\n',
    { tag: 'proto', start: 's', tokenClasses: true })
  if (!probe?.options?.tokenSet?.a) {
    throw new Error('@tabnas/proto: the installed @tabnas/bnf does not ' +
      'support tokenClasses; upgrade @tabnas/bnf (tabnas/bnf#74)')
  }
  // `wordKeywords` makes a keyword match as a whole word; `tokenClasses`
  // compiles `ident` (an identifier or any keyword) to one engine token
  // set, so a lookahead position peeks it as one token rather than one
  // alternate per keyword. Both are required: see AGENTS.md.
  ;(tn.abnf as Function)(grammarText, {
    tag: 'proto',
    start: 'proto',
    wordKeywords: true,
    tokenClasses: true,
  })
  // An aggregate value (`option (f) = { a: 1 };`) is recorded as the text
  // between its braces, which the CST's `src` does not keep: the lexer
  // drops whitespace and comments. This action reads it from the source
  // while the brace tokens are to hand; see ./aggregate.ts.
  tn.rule('constant', (rs: any) => rs.ac(recordAggregate))
}) as {
  (tn: AnyTabnas, options?: Partial<ProtoOptions>): void
  defaults: ProtoOptions
}

Proto.defaults = { version: null, reconcile: true }

// Turn a parsed proto CST into a FileDescriptorProto, resolving the
// version from the file's declaration and the supplied options.
function toDescriptor(cst: any, options?: Partial<ProtoOptions>): FileDescriptorProto {
  const opts: ProtoOptions = { ...Proto.defaults, ...(options || {}) }
  const first = (cst && cst.kids ? cst.kids : []).find((k: any) => k && k.rule)
  const declared =
    first && 'syntaxOrEdition' === first.rule ? declaredVersion(first) : null
  const version = resolveVersion(declared, opts.version, opts.reconcile)
  return buildFile(cst, version)
}

// Convenience: parse a `.proto` source string to a FileDescriptorProto in
// one call. Builds a fresh engine each time; for repeated parsing reuse an
// engine via `const j = new Tabnas().use(Proto)` and call
// `toDescriptor(j.parse(src), opts)`.
function parse(src: string, options?: Partial<ProtoOptions>): FileDescriptorProto {
  const tn = new Tabnas({ rewind: { history: 8192 } }) as AnyTabnas
  Proto(tn, options)
  return toDescriptor(tn.parse(src), options)
}

// VERSION is this package's version. It MUST equal package.json "version":
// the release orchestrator rewrites both, and test/version.test.ts fails the
// build if they drift. Mirrors `const VERSION` in go/proto.go.
const VERSION = '0.6.0'

export { Proto, parse, toDescriptor, VERSION }
export type { ProtoVersion }
export * from './descriptor'

File: go/proto.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// Package tabnasproto parses Protocol Buffers .proto IDL (proto2, proto3,
// edition 2023/2024) into FileDescriptorProto-shaped Go values. It drives the
// Tabnas engine with an ABNF grammar (via @tabnas/abnf) rather than a
// hand-written parser. The version is auto-detected from the file's
// syntax/edition declaration and/or supplied via ProtoOptions.
//
// Go port of ts/src/proto.ts.
package tabnasproto

import (
	abnf "github.com/tabnas/abnf/go"
	tabnas "github.com/tabnas/parser/go"
)

// VERSION is this module's version. It MUST equal ts/package.json
// "version": the release orchestrator rewrites both, and
// TestVersionMatchesPackageJSON fails the build if they drift.
const VERSION = "0.5.1"

//go:generate go run grammar_gen.go

// ProtoOptions configures descriptor construction.
type ProtoOptions struct {
	// Version forces the protobuf version. "" (the default) auto-detects from
	// the file's syntax/edition declaration.
	Version ProtoVersion
	// Reconcile, when nil (the default) or true, errors if an explicit Version
	// disagrees with the file's declaration; when false the declaration wins.
	Reconcile *bool
}

// Proto installs the union proto grammar onto j so it can parse .proto source
// into a {rule, src, kids} CST. Use ToDescriptor to turn that CST into a
// FileDescriptorProto. Mirrors the TS `tn.use(Proto)` plugin.
func Proto(j *tabnas.Tabnas) error {
	// Proto drives parsing through @tabnas/abnf. wordKeywords is required so
	// literal keywords match as whole words (e.g. `option` does not grab the
	// `option` prefix of `optional`).
	_, err := abnf.Install(j, GrammarText, &abnf.AbnfConvertOptions{
		Tag:          "proto",
		Start:        "proto",
		WordKeywords: true,
	}, nil)
	if err != nil {
		return err
	}
	// An aggregate value (`option (f) = { a: 1 };`) is recorded as the text
	// between its braces, which the CST's src does not keep: the lexer drops
	// whitespace and comments. This action reads it from the source while
	// the brace tokens are to hand; see aggregate.go.
	var pushed []string
	j.Rule("constant", func(rs *tabnas.RuleSpec, _ *tabnas.Parser) {
		rs.AddAC(recordAggregate)
		pushed = pushedRules(rs)
	})
	// Every rule inside an aggregate value carries a mark, set on the rules
	// the `constant` rule pushes and kept by the rules below them, so the
	// matchers below can tell inside from outside at once; see aggregate.go.
	for _, name := range pushed {
		j.Rule(name, func(rs *tabnas.RuleSpec, _ *tabnas.Parser) {
			rs.AddBO(markAggregate)
		})
	}
	// Text format has no keywords: inside an aggregate value a word the
	// grammar spells as a keyword, and a bracketed extension or Any name, are
	// identifiers. This matcher runs ahead of the grammar's own (priority
	// 1e6) and reads them so there; see aggregate.go. The second refuses
	// adjacent string literals that protoc's tokenizer refuses; see
	// strings.go.
	j.SetOptions(tabnas.Options{Lex: &tabnas.LexOptions{Match: map[string]*tabnas.MatchSpec{
		"protoAggregateWord":   {Order: 900000, Make: makeAggregateWord},
		"protoAdjacentStrings": {Order: 950000, Make: makeAdjacentStrings},
	}}})
	return nil
}

// ToDescriptor turns a parsed proto CST into a FileDescriptorProto, resolving
// the version from the file's declaration and the supplied options.
func ToDescriptor(cst any, opts *ProtoOptions) (FileDescriptorProto, error) {
	version := ProtoVersion("")
	reconcile := true
	if opts != nil {
		version = opts.Version
		if opts.Reconcile != nil {
			reconcile = *opts.Reconcile
		}
	}

	root, _ := cst.(map[string]any)
	var first map[string]any
	for _, k := range nkids(root) {
		if nrule(k) != "" {
			first = k
			break
		}
	}
	declared := ProtoVersion("")
	if first != nil && nrule(first) == "syntaxOrEdition" {
		d, err := DeclaredVersion(first)
		if err != nil {
			return FileDescriptorProto{}, err
		}
		declared = d
	}

	resolved, err := ResolveVersion(declared, version, reconcile)
	if err != nil {
		return FileDescriptorProto{}, err
	}
	return BuildFile(root, resolved), nil
}

// Parse parses a .proto source string to a FileDescriptorProto in one call.
// It builds a fresh engine each time; for repeated parsing reuse an engine:
// build one with tabnas.Make, install Proto, then call ToDescriptor(j.Parse(src)).
func Parse(src string, opts *ProtoOptions) (FileDescriptorProto, error) {
	rh := 8192
	j := tabnas.Make(tabnas.Options{Rewind: &tabnas.RewindOptions{History: &rh}})
	if err := Proto(j); err != nil {
		return FileDescriptorProto{}, err
	}
	cst, err := j.Parse(src)
	if err != nil {
		return FileDescriptorProto{}, err
	}
	return ToDescriptor(cst, opts)
}
